### ISSUE TITLE
fiber from `ZManaged.useForever` not interrupted when dying on main fiber

### ISSUE THREAD

**Issue by felher** [zio/zio#2968]
Hello folks,
I'm not sure this is really an issue or just a misunderstanding on my part, but here it goes.

Here is a small program, which creates an resource of a blaze http server, waits 3 seconds and then either does nothing, interrupts the server or just dies (Or a combination of the latter two). A link to a full example ready to be cloned is below.

```scala
def run(args: List[String]): ZIO[ZEnv, Nothing, Int] =
  for {
    s  <- server.useForever.fork
    _  <- ZIO.sleep(Duration(3, TimeUnit.SECONDS))
    _  <- if (args.contains("interrupt")) s.interrupt else UIO.unit
    _  <- if (args.contains("die")) UIO.die(new Throwable("end")) else UIO.unit
  } yield 0

def server: ZManaged[Any, Throwable, Server[Task]] =
  ZManaged
    .make(
      BlazeServerBuilder[Task]
        .bindHttp(8745)
        .withHttpApp(helloWorldService.orNotFound)
        .allocated
    )(x =>
      UIO({println("Destroying")}) *>
      x._2.orDie *>
      UIO { println("Destroyed")})
    .map(_._1)

```

When the program is run without arguments, the blaze sever is shut down as expected and the program exits with zero. If, however, the program is run with "die", then the server is not cleaned up and the program never stops. Here is an overview:

|arguments|cleans up and exits|
|--|--|
||yes|
|"interrupt"|yes|
|"interrupt", "die"|yes|
|"die"|no|

I did expect all four paths through the program to end the same way.

The following dependencies were used:
```
  "dev.zio"         %% "zio"                 % "1.0.0-RC17",
  "dev.zio"         %% "zio-interop-cats"    % "2.0.0.0-RC10",
  "org.http4s"      %% "http4s-blaze-server" % "0.21.1",
  "org.http4s"      %% "http4s-dsl"          % "0.21.1",
  "ch.qos.logback"   % "logback-classic"     % "1.2.3",
```

You can see a reproduction here: https://github.com/felher/zio-blaze-reproduction
Clone it and start sbt. The Revolver-Plugin is included, so you can type `reStart` and see the program exit normally after 3 seconds with exit code zero. The same goes for `reStart interrupt` and `reStart interrupt die`. Only `reStart die` renders the program unresponsive.

Here is a example:
```

> git clone https://github.com/felher/zio-blaze-reproduction
> cd zio-blaze-reproduction
> sbt
[info] Loading settings for project global-plugins from metals.sbt ...
...
sbt:server> reStart
[info] Application tmp-w9zwntrjtn not yet started
[info] Starting application tmp-w9zwntrjtn in the background ...
...
tmp-w9zwntrjtn Destroying
...
tmp-w9zwntrjtn Destroyed
tmp-w9zwntrjtn ... finished with exit code 0

sbt:server> reStart interrupt
[info] Application tmp-w9zwntrjtn not yet started
[info] Starting application tmp-w9zwntrjtn in the background ...
...
tmp-w9zwntrjtn Destroying
...
tmp-w9zwntrjtn Destroyed
tmp-w9zwntrjtn ... finished with exit code 0

sbt:server> reStart interrupt die
[info] Application tmp-w9zwntrjtn not yet started
[info] Starting application tmp-w9zwntrjtn in the background ...
...
tmp-w9zwntrjtn Destroying
...
tmp-w9zwntrjtn Destroyed
tmp-w9zwntrjtn[ERROR] Fiber failed.
tmp-w9zwntrjtn[ERROR] An unchecked error was produced.
tmp-w9zwntrjtn[ERROR] java.lang.Throwable: end
tmp-w9zwntrjtn[ERROR]   at org.felher.tmpW9zwNTRjTn.Main$.$anonfun$run$3(Main.scala:26)
...
tmp-w9zwntrjtn ... finished with exit code 1

sbt:server> reStart die
[info] Application tmp-w9zwntrjtn not yet started
[info] Starting application tmp-w9zwntrjtn in the background ...
...
tmp-w9zwntrjtn[ERROR] Fiber failed.
tmp-w9zwntrjtn[ERROR] An unchecked error was produced.
tmp-w9zwntrjtn[ERROR] java.lang.Throwable: end
tmp-w9zwntrjtn[ERROR]   at org.felher.tmpW9zwNTRjTn.Main$.$anonfun$run$3(Main.scala:26)


    ########################
    ## No 'finished with exit code `'
    ## Program is still running
    ## I expected it to finish with 1, too
```

I'm sorry that I was not able to golf it down further. A minimal example like the following works as expected.

```scala
  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] = for {
    _ <- Managed.unit.useForever.fork
    _ <- ZIO.sleep(zio.duration.Duration(3, java.util.concurrent.TimeUnit.SECONDS))
    _ <- ZIO.die(new Throwable("hi"))
  } yield 3
```

**Comment by mlangc** [comment]
You can have exactly the behavior you want by rewriting your `run` function a bit:

```scala
  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] = {
    server.use { _ =>
      for {
        _ <- ZIO.sleep(Duration(3, TimeUnit.SECONDS))
        _ <- if (args.contains("interrupt")) ZIO.interrupt else UIO.unit
        _ <- if (args.contains("die")) UIO.die(new Throwable("end")) else UIO.unit
      } yield 0
    }
  }.orDie
```

I thought about this for a while, and I don't think that the behavior you describe is a bug. `ZManaged` guaranties that resources are closed as soon as the `use` block terminates. By using `useForever` though, this will never happen, unless you interrupt it explicitly - as you do in the `interrupt` case. What confuses you I think, is that killing the parent fiber does not interrupt the child fiber. Note however, that while this makes sense in some situations, it is not desirable in general. It might be worthwhile to support this behavior via a special combinator - similar to `ZIO.deamon`- but I guess that this has already been discussed. @adamgfraser @iravid what is your take on this?

**Comment by iravid** [comment]
In RC17, parent fibers dying will not cause child fibers to be interrupted. Only interrupting the parent fiber will cause the child fiber to be interrupted. So I don't think this is a bug, no. 

This is partly why it's usually advisable to construct the entire application as a `ZManaged` instance, so that finalizers for resources always run. @mlangc's formulation looks pretty good. Here's another take:
```scala
def run(args: List[String]): ZIO[ZEnv, Nothing, Int] =
  (for {
    s  <- server.fork
    _  <- ZIO.sleep(Duration(3, TimeUnit.SECONDS)).toManaged_
    _  <- s.interrupt.when(args.contains("interrupt")).toManaged_
    _  <- ZIO.dieMessage("end").when(args.contains("die"))
  } yield 0).use(UIO.succeed)
```

Note that this uses `ZManaged` for composing the app, and in particular uses `ZManaged#fork` for forking off the server process in the background while ensuring that fiber is interrupted as part of the ZManaged release process.

As a side note, in the upcoming #2748, parent fiber termination (successfully/due to failure) will cause child fiber interruption, so your original example should work as you expect.

**Comment by adamgfraser** [comment]
Resolved by #2748.

**Comment by felher** [comment]
Ah, great. Thanks to @mlangc  and @iravid for the explanation why this isn't a bug and thanks to @adamgfraser and @jdegoes for fixing it nonetheless.
